File: Webvantage/Webvantage/src/models/employee/expense-report.ts
import { ModelBase } from 'models/base/model-base';

export namespace EmployeeExpenseReport {

    export class ExpenseReportDetail extends ModelBase {
        private _itemDate: Date

        ID: number;
        InvoiceNumber: number;
        LineNumber: number;

        get ItemDate(): Date {
            return this._itemDate;
        }
        set ItemDate(value: Date) {
            this._itemDate = this.getShortDate(value);
        }

        Description: string;
        CreditCardFlag: boolean;
        ClientCode: string;
        DivisionCode: string;
        ProductCode: string;
        JobNumber: number;
        JobDescription: string;
        JobComponentNumber: number;
        JobComponentDescription: string;
        FunctionCode: string;
        Quantity: number;
        Rate: number;
        CreditCardAmount: number;
        Amount: number;
        APComment: string;
        CreatedBy: string;
        ModifiedBy: string;
        ModifiedDate: Date;
        PaymentType: number;
        IsImported: boolean;
        NonBillable: boolean;


        constructor() {
            super();
        }

        copyInto(e) {

            this.ID = e.ID;
            this.InvoiceNumber = e.InvoiceNumber;
            this.LineNumber = e.LineNumber;
            this.ItemDate = e.ItemDate;
            this.Amount = e.Amount;
            this.APComment = e.APComment;
            this.ClientCode = e.ClientCode;
            this.CreditCardFlag = e.CreditCardFlag;
            this.Description = e.Description;
            this.DivisionCode = e.DivisionCode;
            this.ProductCode = e.ProductCode;
            this.JobNumber = e.JobNumber;
            this.JobDescription = e.JobDescription;
            this.JobComponentNumber = e.JobComponentNumber;
            this.JobComponentDescription = e.JobComponentDescription;
            this.FunctionCode = e.FunctionCode;
            this.Quantity = e.Quantity;
            this.Rate = e.Rate;
            this.CreditCardAmount = e.CreditCardAmount;
            this.CreatedBy = e.CreatedBy;
            this.ModifiedBy = e.ModifiedBy;
            this.ModifiedDate = e.ModifiedDate;
            this.PaymentType = e.PaymentType;
            this.IsImported = e.IsImported;

        }

    }

    export class ExpenseReport extends ModelBase {

        ApprovedBy: string;
        ApprovedDate: Date;
        ApproverNotes: string;
        BatchDate: Date;
        CreatedBy: string;
        CreatedDate: Date;
        DateFrom: Date;
        DateTo: Date;
        Description: string;
        Details: string;
        EmployeeCode: string;
        EmployeeName: string;
        InvoiceAmount: number;
        InvoiceDate: Date;
        InvoiceNumber: number;
        IsApproved: boolean;
        IsSubmitted: boolean;
        ModifiedBy: string;
        ModifiedDate: Date;
        Status: number;
        StatusCode: string;
        SubmittedTo: string;
        VendorCode: string;
        TotalNonBillable: number;
        TotalBillable: number;
        TotalAmount: number;
        Paid: boolean;

        ExpenseReportDetails: ExpenseReportDetail[];

        constructor() {
            super();
        }

    }

    export enum ExpenseReportStatus {
        Open = 0,
        Pending = 1,
        Approved = 2,
        PendingApproval = 3,
        DeniedByApprover = 4,
        ApprovedByApprover = 5,
        ApprovedInAccounting = 6,
        PendingApprovalInAccounting = 7,
        DeniedByAccounting = 8
    }

    export enum ApprovedFlag {
        Pending = 0,
        Denied = 1,
        Approved = 2
    }
}

File: AdvantageFramework.BLogic/DTO/Employee/ExpenseReport.vb
Namespace DTO.Employee

    Public Class ExpenseReport

        Public Property ApprovedBy As String
        Public Property ApprovedDate As Date?
        Public Property ApproverNotes As String
        Public Property BatchDate As Date?
        Public Property CreatedBy As String
        Public Property CreatedDate As Date?
        Public Property DateFrom As Date?
        Public Property DateTo As Date?
        Public Property Description As String
        Public Property Details As String
        Public Property EmployeeCode As String
        Public Property EmployeeName As String
        Public Property InvoiceAmount As Decimal?
        Public Property InvoiceDate As Date?
        Public Property InvoiceNumber As Integer
        Public Property IsApproved As Short?
        Public Property IsSubmitted As Short?
        Public Property ModifiedBy As String
        Public Property ModifiedDate As Date?
        Public Property Status As Integer?
        Public ReadOnly Property StatusCode As String

            Get
                If Not Status.HasValue Then
                    Return ""
                Else
                    Return [Enum].GetName(GetType(AdvantageFramework.ExpenseReports.ExpenseReportStatus), AdvantageFramework.ExpenseReports.LoadExpenseReportStatus(Status, IsSubmitted, IsApproved)) '[Enum].GetName(GetType(AdvantageFramework.ExpenseReports.ExpenseReportStatus), CType(Status, Int32))
                End If

            End Get

        End Property
        Public Property SubmittedTo As String
        Public Property VendorCode As String
        Public Property TotalNonBillable As Decimal?
        Public Property TotalBillable As Decimal?
        Public Property TotalAmount As Decimal?
        Public Property Paid As Boolean
        Public Property ExpenseReportDetails As Generic.List(Of ExpenseReportDetail)

        Public Sub New()

        End Sub

        Public Function ToEntity(ByVal DbContext As AdvantageFramework.Database.DbContext) As AdvantageFramework.Database.Entities.ExpenseReport

            'objects
            Dim ExpenseReportEntity As AdvantageFramework.Database.Entities.ExpenseReport = Nothing

            If DbContext IsNot Nothing Then

                ExpenseReportEntity = AdvantageFramework.Database.Procedures.ExpenseReport.LoadByInvoiceNumber(DbContext, Me.InvoiceNumber)
            End If

            If ExpenseReportEntity Is Nothing Then

                ExpenseReportEntity = New Database.Entities.ExpenseReport

            End If

            With ExpenseReportEntity
                .ApprovedBy = Me.ApprovedBy
                .ApprovedDate = Me.ApprovedDate
                .ApproverNotes = Me.ApproverNotes
                .BatchDate = Me.BatchDate
                .CreatedBy = Me.CreatedBy
                .CreatedDate = Me.CreatedDate
                .DateFrom = Me.DateFrom
                .DateTo = Me.DateTo
                .Description = Me.Description
                .Details = Me.Details
                .EmployeeCode = Me.EmployeeCode
                .InvoiceAmount = Me.InvoiceAmount
                .InvoiceDate = Me.InvoiceDate
                .InvoiceNumber = Me.InvoiceNumber
                .IsApproved = Me.IsApproved
                .IsSubmitted = Me.IsSubmitted
                .ModifiedBy = Me.ModifiedBy
                .ModifiedDate = Me.ModifiedDate
                .Status = Me.Status
                .ApproverNotes = Me.ApproverNotes
                .IsSubmitted = Me.IsSubmitted
                .SubmittedTo = Me.SubmittedTo
                .IsApproved = Me.IsApproved
                .VendorCode = Me.VendorCode

                '.TotalNonBillable = Me.TotalNonBillable
                '.TotalBillable = Me.TotalBillable
                '.TotalAmount = Me.TotalAmount
                '.Paid = Me.Paid

            End With


            Return ExpenseReportEntity

        End Function
        Public Shared Function FromEntity(ByVal Entity As AdvantageFramework.Database.Entities.ExpenseReport) As ExpenseReport

            'objects
            Dim ExpenseReport As AdvantageFramework.DTO.Employee.ExpenseReport = Nothing

            ExpenseReport = New AdvantageFramework.DTO.Employee.ExpenseReport

            With ExpenseReport

                .ApprovedBy = Entity.ApprovedBy
                .ApprovedDate = Entity.ApprovedDate
                .ApproverNotes = Entity.ApproverNotes
                .BatchDate = Entity.BatchDate
                .CreatedBy = Entity.CreatedBy
                .CreatedDate = Entity.CreatedDate
                .DateFrom = Entity.DateFrom
                .DateTo = Entity.DateTo
                .Description = Entity.Description
                .Details = Entity.Details
                .EmployeeCode = Entity.EmployeeCode
                .InvoiceAmount = Entity.InvoiceAmount
                .InvoiceDate = Entity.InvoiceDate
                .InvoiceNumber = Entity.InvoiceNumber
                .IsApproved = Entity.IsApproved
                .IsSubmitted = Entity.IsSubmitted
                .ModifiedBy = Entity.ModifiedBy
                .ModifiedDate = Entity.ModifiedDate
                .Status = Entity.Status
                .SubmittedTo = Entity.SubmittedTo
                .VendorCode = Entity.VendorCode
                '.TotalNonBillable = Entity.TotalNonBillable
                '.TotalBillable = Entity.TotalBillable
                '.TotalAmount = Entity.TotalAmount
                '.Paid = Entity.Paid
            End With

            Return ExpenseReport

        End Function

        Public Shared Function FromClass(ByVal Entity As AdvantageFramework.Database.Classes.ExpenseReport) As ExpenseReport

            'objects
            Dim ExpenseReport As AdvantageFramework.DTO.Employee.ExpenseReport = Nothing

            ExpenseReport = New AdvantageFramework.DTO.Employee.ExpenseReport

            With ExpenseReport

                .ApprovedBy = Entity.ApprovedBy
                .ApprovedDate = Entity.ApprovedDate
                .ApproverNotes = Entity.ApproverNotes
                '.BatchDate = Entity.BatchDate
                '.CreatedBy = Entity.CreatedBy
                .CreatedDate = Entity.CreatedDate
                .DateFrom = Entity.DateFrom
                .DateTo = Entity.DateTo
                .Description = Entity.Description
                .Details = Entity.Details
                .EmployeeCode = Entity.EmployeeCode
                .InvoiceAmount = Entity.InvoiceAmount
                .InvoiceDate = Entity.InvoiceDate
                .InvoiceNumber = Entity.InvoiceNumber
                .IsApproved = Entity.IsApproved
                .IsSubmitted = Entity.IsSubmitted
                '.ModifiedBy = Entity.ModifiedBy
                '.ModifiedDate = Entity.ModifiedDate
                .Status = Entity.Status
                .SubmittedTo = Entity.SubmittedTo
                .VendorCode = Entity.VendorCode
                .TotalNonBillable = Entity.TotalNonBillable
                .TotalBillable = Entity.TotalBillable
                .TotalAmount = Entity.TotalAmount
                .Paid = Entity.Paid
            End With

            Return ExpenseReport


        End Function

    End Class

End Namespace


